### ISSUE TITLE
Fit modes give different results

### ISSUE THREAD

**Issue by oscarkey** [PriorLabs/TabPFN#631]
### Describe the bug

Different settings of the `fit_mode` option give different outputs. The fit mode shouldn't affect the output, up to float precision issues.

### Steps/Code to Reproduce

```python
import random
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
import numpy as np
import torch

from tabpfn import TabPFNRegressor

X, y = load_diabetes(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.33, random_state=42
)


def _set_seeds() -> None:
    torch.manual_seed(0)
    np.random.seed(0)
    random.seed(0)


_set_seeds()
reg = TabPFNRegressor(fit_mode="low_memory", inference_precision=torch.float64)
reg.fit(X_train, y_train)
preds_no_cache = reg.predict(X_test)

reg = TabPFNRegressor(fit_mode="low_memory", inference_precision=torch.float64)
reg.fit(X_train, y_train)
preds_no_cache_repeat = reg.predict(X_test)

_set_seeds()
reg = TabPFNRegressor(fit_mode="fit_preprocessors", inference_precision=torch.float64)
reg.fit(X_train, y_train)
preds_cache_preproc = reg.predict(X_test)

_set_seeds()
reg = TabPFNRegressor(fit_mode="fit_with_cache", inference_precision=torch.float64)
reg.fit(X_train, y_train)
preds_kv_cache = reg.predict(X_test)


def _max_diff(a: np.ndarray, b: np.ndarray) -> float:
    return np.max(np.abs(a - b) / np.abs(a))


print("max relative diffs")
print("no_cache vs no_cache_repeat:", _max_diff(preds_no_cache, preds_no_cache_repeat))
print("no_cache vs cache_preproc:", _max_diff(preds_no_cache, preds_cache_preproc))
print("no_cache vs kv_cache:", _max_diff(preds_no_cache, preds_kv_cache))
```

I set `TABPFN_EXCLUDE_DEVICES=mps` to get float64 support on MacOS.

### Expected Results

```
no_cache == no_cache_repeat: 0.0
no_cache == cache_preproc: ~0.0
no_cache == kv_cache: ~0.0
```

### Actual Results

```
no_cache == no_cache_repeat: True
no_cache == cache_preproc: 0.044
no_cache == kv_cache: 1.423
```

### Versions

```shell
PyTorch version: 2.9.0
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.4.4.1)
CMake version: version 4.1.2
Libc version: N/A

Python version: 3.10.17 (main, Apr  9 2025, 03:47:39) [Clang 20.1.0 ] (64-bit runtime)
Python platform: macOS-26.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M4

Dependency Versions:
--------------------
tabpfn: 6.0.6
torch: 2.9.0
numpy: 2.2.6
scipy: 1.15.3
pandas: 2.3.3
scikit-learn: 1.7.2
typing_extensions: 4.15.0
einops: 0.8.1
huggingface-hub: 1.1.2
```

**Comment by oscarkey** [comment]
Fixed by Magnus above!

**Comment by oscarkey** [comment]
oh no wait the pr isn't merged yet...

**Comment by noahho** [comment]
@magnus-priorlabs @MagnusBuehler did you ever end up merging the fix?

**Comment by MagnusBuehler** [comment]
@noahho  This was on my backburner for quite some time but I never finished it. I created a PR (not yet done) with part of the solution here: https://github.com/PriorLabs/TabPFN/pull/784

**Comment by noahho** [comment]
okay great, just making sure the work doesnt get lost! Also worth checking if this is now fixed with the recent preprocerssing refactor?

**Comment by klemens-floege** [comment]
What is the current status of this? Can we close the Issue or do we still need to merge the fix?

**Comment by noahho** [comment]
Yeah would be great to wrap up! @magnus-priorlabs when would you find time for this?

**Comment by egeonur** [comment]
hey everyone I am Ege from LMU Data Science Master and became fan of your project I tried to fix this issue and opened this pr #802 I tried to follow your code format and tried to do as minimal changes as possbile to fix it. Appreciate if you can take a look into that and give feedback thanks.

**Comment by eddiebergman** [comment]
Seems a lot better than it was before, but still some slight inconsistencies between no-cache and kv-cache.

```bash
$ TABPFN_EXCLUDE_DEVICES=mps uv run t.py

max relative diffs
no_cache vs no_cache_repeat: 0.0
no_cache vs cache_preproc: 0.0
no_cache vs kv_cache: 0.0002758397
```

---

```
Python 3.13.11
Apple M4

$ uv run pip list
annotated-doc              0.0.4
annotated-types            0.7.0
anyio                      4.14.0
ast_serialize              0.5.0
attrs                      26.1.0
babel                      2.18.0
backrefs                   7.0
black                      26.5.1
boolean.py                 5.0
cattrs                     26.1.0
certifi                    2026.6.17
cfgv                       3.5.0
charset-normalizer         3.4.7
click                      8.4.1
colorama                   0.4.6
configurator               3.2.0
depgather                  0.5.0
distlib                    0.4.3
einops                     0.8.2
execnet                    2.1.2
filelock                   3.29.4
fsspec                     2026.6.0
ghp-import                 2.1.0
griffelib                  2.1.0
h11                        0.16.0
hf-xet                     1.5.1
httpcore                   1.0.9
httpx                      0.28.1
huggingface_hub            1.20.1
identify                   2.6.19
idna                       3.18
iniconfig                  2.3.0
Jinja2                     3.1.6
joblib                     1.5.3
librt                      0.11.0
license-expression         30.4.4
licensecheck               2026.0.8
lightgbm                   4.6.0
loguru                     0.7.3
Markdown                   3.10.2
markdown-exec              1.12.1
markdown-it-py             4.2.0
MarkupSafe                 3.0.3
mdurl                      0.1.2
mergedeep                  1.3.4
mike                       2.2.0
mkdocs                     1.6.1
mkdocs-autorefs            1.4.4
mkdocs-gen-files           0.6.1
mkdocs-get-deps            0.2.2
mkdocs-glightbox           0.5.2
mkdocs-literate-nav        0.6.3
mkdocs-material            9.7.6
mkdocs-material-extensions 1.3.1
mkdocstrings               1.0.4
mkdocstrings-python        2.0.5
ml_dtypes                  0.5.4
mlx                        0.31.2
mlx-metal                  0.31.2
mpmath                     1.3.0
mypy                       2.1.0
mypy_extensions            1.1.0
narwhals                   2.22.1
networkx                   3.6.1
nodeenv                    1.10.0
numpy                      2.4.6
onnx                       1.22.0
packaging                  26.2
paginate                   0.5.7
pandas                     3.0.3
pathspec                   1.1.1
pip                        26.1.2
platformdirs               4.10.0
pluggy                     1.6.0
pre_commit                 4.6.0
properdocs                 1.6.7
protobuf                   7.35.1
pydantic                   2.13.4
pydantic_core              2.46.4
pydantic-settings          2.14.2
Pygments                   2.20.0
pygments-ansi-color        0.3.0
pymdown-extensions         10.21.3
pyparsing                  3.3.2
pytest                     9.1.1
pytest-mock                3.15.1
pytest-xdist               3.8.0
python-dateutil            2.9.0.post0
python-discovery           1.4.2
python-dotenv              1.2.2
pytokens                   0.4.1
PyYAML                     6.0.3
pyyaml_env_tag             1.1
requests                   2.34.2
requests-cache             1.3.2
requirements-parser        0.13.1
rich                       15.0.0
ruff                       0.15.12
safetensors                0.8.0
scikit-learn               1.9.0
scipy                      1.18.0
selectolax                 0.4.10
setuptools                 81.0.0
shellingham                1.5.4
six                        1.17.0
sympy                      1.14.0
tabpfn                     8.0.8
threadpoolctl              3.6.0
tomli                      2.4.1
torch                      2.12.1
towncrier                  25.8.0
tqdm                       4.68.3
typer                      0.25.1
typing_extensions          4.15.0
typing-inspection          0.4.2
url-normalize              3.0.0
urllib3                    2.7.0
verspec                    0.1.0
virtualenv                 21.5.1
watchdog                   6.0.0
```

**Comment by oscarkey** [comment]
I got claude to look into this, and the cause is the kv cache being quantised to 8 bits. I'll add this to the pydoc.
Might also be nice to disable quantisation if the user specifies an inference precision. @bejaeger do you have a vision for how this should work?

**Comment by bejaeger** [comment]
Yes, I agree that we should disable quantization if an inference_precision is specified!
Also, we should document that the default kv_cache uses quantization and that numerical differences are expected due to that, but no performance loss. We can do that in the inference engine class

**Comment by klemens-floege** [comment]
@jmkuebler the PR that you linked was merged, can we close this issue now or would you rather leave it open?

**Comment by jmkuebler** [comment]
Yes let's close this 👍
